src/clients/UBAClient/UBAClientWithRefresh.ts

```diff
@@ -2,165 +2,1122 @@ import assert from "assert";
+   * @param flowChain
+   * @returns
+   */
+  private getUbaBundleBlockRangeContainingFlow(flowBlock: number, flowChain: number): number[][] {
+    const blockRangesContainingFlow = this.ubaBundleBlockRanges.find((blockRanges) => {
+      const blockRangeForChain = getBlockRangeForChain(blockRanges, flowChain, this.chainIdIndices);
+      return blockRangeForChain[0] <= flowBlock && flowBlock <= blockRangeForChain[1];
+    });
+    if (!blockRangesContainingFlow) {
+      throw new Error(`Could not find bundle block range containing flow at block ${flowBlock} on chain ${flowChain}`);
+    }
+    return blockRangesContainingFlow;
+  }
+
+  /**
+   * This client should only update the `ubaBundleStates` state via this function.
+   */
+  private appendValidatedFlowsToClassState(
+    chainId: number,
+    tokenSymbol: string,
+    flows: ModifiedUBAFlow[],
+    bundleBlockRanges: number[][]
+  ) {
+    const bundleState = this.getBundleState(bundleBlockRanges, tokenSymbol, chainId);
+    bundleState.flows.push(...flows);
+  }
+
+  /**
+   * Updates the bundle state.
+   */
+  public async update(): Promise<void> {
+    const latestHubPoolBlock = this.hubPoolClient.latestBlockNumber;
+    if (!isDefined(latestHubPoolBlock)) throw new Error("HubPoolClient not updated");
+
+    this.logger.debug({
+      at: "UBAClientWithRefresh",
+      message: "❤️‍🔥😭  Updating UBA Client",
+    });
+
+    const tokens = this.tokens;
+
+    // Load all UBA bundle block ranges for each chain:
+    const blockRangesByChain = this.getMostRecentBundleBlockRangesPerChain(100);
+    const bundleBlockRangesCount = blockRangesByChain[this.chainIdIndices[0]].length;
+    const bundleBlockRanges: number[][][] = [];
+    for (let i = 0; i < bundleBlockRangesCount; i++) {
+      bundleBlockRanges.push(
+        this.chainIdIndices.map((chainId) => {
+          return blockRangesByChain[chainId][i];
+        })
+      );
+    }
+    this.ubaBundleBlockRanges = bundleBlockRanges;
+    console.log("UBA bundle block ranges we're storing in class memory", this.ubaBundleBlockRanges);
+
+    // Pre-load bundle configs and opening balances for each chain and token per bundle range.
+    // This also instantiates the ubaBundleStates class dictionary.
+    for (let i = 0; i < this.ubaBundleBlockRanges.length; i++) {
+      const bundleBlockRange = this.ubaBundleBlockRanges[i];
+      const startBlocks = bundleBlockRange.map(([start]) => start);
+      const mainnetStartBlock = getBlockForChain(startBlocks, this.hubPoolClient.chainId, this.chainIdIndices);
+      tokens.forEach((token) => {
+        const l1TokenAddress = this.hubPoolClient.getL1Tokens().find((_token) => _token.symbol === token)?.address;
+        if (!isDefined(l1TokenAddress)) throw new Error(`No L1 token address found for token symbol ${token}`);
+        this.chainIdIndices.forEach((chainId) => {
+          // if (!isDefined(this.spokePoolClients[chainId])) {
+          //   return;
+          // }
```

Comment:
These should _always_ be defined, no?

Reply:
well, except for the boba one it won't be but we don't need to check that here.

Reply:
Ah, correct - forgot that this should also include disabled chains.